src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,161 @@
+            jobContainer.ContainerId = response.Context.Container.Id;
+            context.JobContext.Container["network"] = new StringContextData(response.Context.Container.Network);
+            jobContainer.ContainerNetwork = response.Context.Container.Network;
+            context.JobContext["hook_state"] = new StringContextData(JsonUtility.ToString(response.State));
+            // var configEnvFormat = "--format \"{{range .Config.Env}}{{println .}}{{end}}\"";
+            // var containerEnv = await _dockerManager.DockerInspect(executionContext, container.ContainerId, configEnvFormat);
+            // container.ContainerRuntimePath = DockerUtil.ParsePathFromConfigEnv(containerEnv);
+
+            for (var i = 0; i < response.Context.Services.Count; i++)
+            {
+                var container = response.Context.Services[i]; // TODO: Confirm that the order response.Context.Services is the same as serviceContainers
+                var containerInfo = serviceContainers[i];
+                containerInfo.ContainerId = container.Id;
+                containerInfo.ContainerNetwork = container.Network;
+                var service = new DictionaryContextData()
+                {
+                    ["id"] = new StringContextData(container.Id),
+                    ["ports"] = new DictionaryContextData(),
+                    ["network"] = new StringContextData(container.Network)
+                };
+
+                // TODO: workout port mappings + format
+                // foreach (var portMapping in containerInfo.UserPortMappings)
+                // {
+                //     // TODO: currently the format is ports["80:8080"] = "80:8080", fix this?
+                //     (service["ports"] as DictionaryContextData)[$"{portMapping.Key}:{portMapping.Value}"] = new StringContextData($"{portMapping.Key}:{portMapping.Value}");
+                // }
+                context.JobContext.Services[containerInfo.ContainerNetworkAlias] = service;
+            }
+        }
+
+        public async Task CleanupJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+
+            var responsePath = GenerateResponsePath();
+            var input = new HookInput
+            {
+                Command = HookCommand.CleanupJob,
+                ResponseFile = responsePath,
+                Args = new HookArgs
+                {
+                    JobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault().GetHookContainer(),
+                    Services = containers.Where(c => c.IsJobContainer == false).Select(c => c.GetHookContainer()).ToList(),
+                    Network = containers.Where(c => !string.IsNullOrEmpty(c.ContainerNetwork)).FirstOrDefault()?.ContainerNetwork,
+                },
+                State = JsonUtility.FromString<dynamic>(context.JobContext["hook_state"].ToString())                
```

Comment:
This will throw an exception if state doesn't get set, i think we need to check if the dictionary contains the key first